pkg/operator/ceph/csi/spec.go

```diff
@@ -561,58 +558,56 @@ func (r *ReconcileCSI) startDrivers(ver *version.Info, ownerInfo *k8sutil.OwnerI
 	return nil
 }
 
-func (r *ReconcileCSI) stopDrivers(ver *version.Info) {
+func (r *ReconcileCSI) stopDrivers(ver *version.Info) error {
+	RBDDriverName = fmt.Sprintf("%s.rbd.csi.ceph.com", r.opConfig.OperatorNamespace)
+	CephFSDriverName = fmt.Sprintf("%s.cephfs.csi.ceph.com", r.opConfig.OperatorNamespace)
+
 	if !EnableRBD {
 		logger.Info("CSI Ceph RBD driver disabled")
-		succeeded := r.deleteCSIDriverResources(ver, csiRBDPlugin, csiRBDProvisioner, "csi-rbdplugin-metrics", RBDDriverName)
-		if succeeded {
-			logger.Info("successfully removed CSI Ceph RBD driver")
-		} else {
-			logger.Error("failed to remove CSI Ceph RBD driver")
+		err := r.deleteCSIDriverResources(ver, csiRBDPlugin, csiRBDProvisioner, "csi-rbdplugin-metrics", RBDDriverName)
+		if err != nil {
+			return errors.Wrap(err, "failed to remove CSI Ceph RBD driver")
 		}
+		logger.Info("successfully removed CSI Ceph RBD driver")
 	}
 
 	if !EnableCephFS {
 		logger.Info("CSI CephFS driver disabled")
-		succeeded := r.deleteCSIDriverResources(ver, csiCephFSPlugin, csiCephFSProvisioner, "csi-cephfsplugin-metrics", CephFSDriverName)
-		if succeeded {
-			logger.Info("successfully removed CSI CephFS driver")
-		} else {
-			logger.Error("failed to remove CSI CephFS driver")
+		err := r.deleteCSIDriverResources(ver, csiCephFSPlugin, csiCephFSProvisioner, "csi-cephfsplugin-metrics", CephFSDriverName)
+		if err != nil {
+			return errors.Wrap(err, "failed to remove CSI CephFS driver")
 		}
+		logger.Info("successfully removed CSI CephFS driver")
 	}
+
+	return nil
 }
 
-func (r *ReconcileCSI) deleteCSIDriverResources(ver *version.Info, daemonset, deployment, service, driverName string) bool {
-	succeeded := true
+func (r *ReconcileCSI) deleteCSIDriverResources(ver *version.Info, daemonset, deployment, service, driverName string) error {
 	csiDriverobj = beta1CsiDriver{}
 	if ver.Major > KubeMinMajor || ver.Major == KubeMinMajor && ver.Minor >= kubeMinVerForV1csiDriver {
 		csiDriverobj = v1CsiDriver{}
 	}
 	err := k8sutil.DeleteDaemonset(r.opManagerContext, r.context.Clientset, r.opConfig.OperatorNamespace, daemonset)
 	if err != nil {
-		logger.Errorf("failed to delete the %q. %v", daemonset, err)
-		succeeded = false
+		return errors.Wrapf(err, "failed to delete the %q", daemonset)
```

Comment:
The good thing about the previous approach is that deleting the csi driver will attempt to delete all of the resources around the csi driver, even if one of the resources failed to be deleted.

The other side effect of this change is that if some resource fails during deletion of the csi driver, the reconcile will fail and it may keep retrying indefinitely. Is it really important to retry though? I'd vote for the previous approach of treating it as best effort, logging any errors, and not failing the reconcile.

Reply:
>The other side effect of this change is that if some resource fails during deletion of the csi driver, the reconcile will fail and it may keep retrying indefinitely. Is it really important to retry though? I'd vote for the previous approach of treating it as best effort, logging any errors, and not failing the reconcile.

if we don't retry we end up having the same problem as mentioned at https://github.com/rook/rook/issues/9693.

Reply:
> The other side effect of this change is that if some resource fails during deletion of the csi driver, the reconcile will fail and it may keep retrying indefinitely. Is it really important to retry though? I'd vote for the previous approach of treating it as best effort, logging any errors, and not failing the reconcile.

>the reconcile will fail and it may keep retrying indefinitely.

ExponentialBackOff will kick in, time period between each retry will increase and utmost we will be trying once every 16.66 minutes within few retries.
I think it's acceptable to retry atleast once in 16 minutes ?

https://github.com/kubernetes/client-go/blob/6964e37acfd3678087cfb1ced00774753c27ea96/util/workqueue/default_rate_limiters.go#L37

Reply:
> Are there any expected conditions where the removal would fail? None of these resources being removed have finalizers that would prevent them from being removed, right? So any deletion failures would just be random api failures and they would be likely to succeed on retry? In that case the retry will be fine. It should be a corner case anyway.

Yes,  any failure would be temporary and deletion will succeed on retry.

Reply:
forgot to mention. noting here for reference as we are using deleting we don't have a problem with the kernel clients as they work even if the csidriver is deleted (but applications cannot be terminated smoothly as no csidriver is present and might leave stale resources on nodes). when userspace clients are used, deleting the csidrivers drops the mounts and applications cannot use the mounts and it cannot be terminated smootly and it can also leave stale resources on the node.

Reply:
> forgot to mention. noting here for reference as we are using deleting we don't have a problem with the kernel clients as they work even if the csidriver is deleted (but applications cannot be terminated smoothly as no csidriver is present and might leave stale resources on nodes). when userspace clients are used, deleting the csidrivers drops the mounts and applications cannot use the mounts and it cannot be terminated smootly and it can also leave stale resources on the node.

Csi drivers will be deleted only when there are zero ceph clusters.

Reply:
> forgot to mention. noting here for reference as we are using deleting we don't have a problem with the kernel clients as they work even if the csidriver is deleted (but applications cannot be terminated smoothly as no csidriver is present and might leave stale resources on nodes). when userspace clients are used, deleting the csidrivers drops the mounts and applications cannot use the mounts and it cannot be terminated smootly and it can also leave stale resources on the node.

Discussed this with @Madhu-1 , 
With internal rook created Ceph cluster, there will no cluster for the volumes to talk to.

most of these concerns are valid only with external ceph clusters and when user deletes it with `Spec.CleanupPolicy.AllowUninstallWithVolumes: true` deliberately. 

Kernel mounted PVC will be unaffected,
Userspace mounted PVC will hand(same as the reboot case)

 Deletion of pod will be difficult. 
Kernel mounted PVC will work as normal once external ceph cluster CR is recreated and csi pods come back online. 
Userspace mounter problem exists even for just csi plugin pod reboot case.

IMO, I don't think we should too concerned about such a corner case where deliberate deletion is involved.

Reply:
Ok thanks for the discussion on the corner case of failing and retrying, makes sense to me.